Automattic/wp-calypso
# composite-checkout: Clean up CompositeCheckout component (3)

#### Changes proposed in this Pull Request

Over time as we've added more features to the implementation of the `composite-checkout` package in calypso, the calypso `CompositeCheckout` component has ended up with quite a lot of code. As of this writing, it is a function around 1200 lines long, and its file contains many helper functions. This PR is an attempt to move code out of the component and into a new directory, `client/my-sites/checkout/composite-checkout/` to make the component easier to read and to reason about going forward.

In fact, this PR moves all of the composite checkout code (and tests) in calypso into this new directory, including the `CompositeCheckout` component itself (which was previously in the confusingly named `checkout/checkout` directory). It also moves the `CheckoutSystemDecider` which is the parent of `CompositeCheckout` into `client/my-sites/checkout` (one level up) which is conceptually where it belongs (one level above both checkout implementations).

Unit tests for much of the composite checkout behavior were added in #40676 which should give additional confidence in this refactor.

#### Testing instructions

This PR shouldn't change any functionality, but it touches almost every part of using composite checkout, so there's a bunch of things that need to be tested.

Here's some things to check:

- Visit composite checkout with a plan in the cart, and verify that the payment methods display correctly (at least credit card and paypal) and that the contact form displays correctly (with just a plan you should see only country and postal code).
- Repeat the above step but with a domain in the cart as well.
- Add a 100% discount coupon to your cart and verify that you see the "free" payment method.
- Cause a Stripe error by using a test card that returns an error and verify that the error is displayed.
- Verify that Tracks analytics are sent to the server on events (like cart load).
- With a domain in the cart, enter invalid data for postal code (like special characters) and press "Continue"; verify that the step does not change and that an error is displayed.
- Complete a payment using the credit card payment method and verify that the thank-you page is displayed.
